### PR TITLE
Inline diagnostics experiment

### PR DESCRIPTION
**Pull request by akhera99**
This is the work done to add the experiment service to inline diagnostics. 
It modifies how the AbstractDiagnosticsTaggerProvider handled the services being enabled. 
Before, it was a property, it was changed to an abstract method. It looks at what the value it set to in VS and accordingly decides whether to retrieve the diagnostics or not.

### CODE REVIEW

File: src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTaggerProvider.cs [hunk 1]
```diff
@@ -31,7 +32,8 @@ namespace Microsoft.CodeAnalysis.Editor.InlineDiagnostics
     internal class InlineDiagnosticsTaggerProvider : AbstractDiagnosticsAdornmentTaggerProvider<InlineDiagnosticsTag>
     {
         private readonly IEditorFormatMap _editorFormatMap;
-        protected sealed override IEnumerable<PerLanguageOption2<bool>> PerLanguageOptions => SpecializedCollections.SingletonEnumerable(InlineDiagnosticsOptions.EnableInlineDiagnostics);
+        //protected sealed override IEnumerable<PerLanguageOption2<bool?>> PerLanguageOptions => SpecializedCollections.SingletonEnumerable(InlineDiagnosticsOptions.EnableInlineDiagnostics);
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTaggerProvider.cs:35]: I guess this should be removed?

**Review by Cosifne — APPROVED** [review]

File: src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTaggerProvider.cs (hunk 1 above)
Comment (akhera99) [reply src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTaggerProvider.cs:35]: ye I forgot, thanks

File: src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTaggerProvider.cs [hunk 2]
```diff
@@ -62,6 +64,20 @@ protected internal override bool IncludeDiagnostic(DiagnosticData diagnostic)
...
         }
 
+        protected internal override bool IsEnabled(Document document)
+        {
+            var workspace = document.Project.Solution.Workspace;
+            var option = workspace.Options.GetOption(InlineDiagnosticsOptions.EnableInlineDiagnostics, document.Project.Language);
+            if (_experimentEnabled is null)
+            {
+                var experimentationService = document.Project.Solution.Workspace.Services.GetRequiredService<IExperimentationService>();
+                _experimentEnabled = experimentationService.IsExperimentEnabled(WellKnownExperimentNames.InlineDiagnostics);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTaggerProvider.cs:74]: i don't get why you need to cache this value.

File: src/VisualStudio/VisualBasic/Impl/Options/AdvancedOptionPageControl.xaml.vb [hunk 3]
```diff
@@ -46,7 +46,15 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.Options
...
             BindToOption(Background_analysis_scope_full_solution, SolutionCrawlerOptions.BackgroundAnalysisScopeOption, BackgroundAnalysisScope.FullSolution, LanguageNames.VisualBasic)
-            BindToOption(DisplayDiagnosticsInline, InlineDiagnosticsOptions.EnableInlineDiagnostics, LanguageNames.VisualBasic)
+            BindToOption(DisplayDiagnosticsInline, InlineDiagnosticsOptions.EnableInlineDiagnostics, LanguageNames.VisualBasic,
+                         Function()
+                             If experimentationService Is Nothing Then
+                                 Return False
+                             End If
+
+                             Return experimentationService.IsExperimentEnabled(WellKnownExperimentNames.InlineDiagnostics)
+                         End Function)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/VisualBasic/Impl/Options/AdvancedOptionPageControl.xaml.vb:56]:
```suggestion
                         Function() If(experimentationService?.IsExperimentEnabled(WellKnownExperimentNames.InlineDiagnostics), False))
```

File: src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTaggerProvider.cs (hunk 2 above)
Comment (akhera99) [reply src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTaggerProvider.cs:74]: Would I want to get the experimentation service every time produced tags is called? This way it doesn't need to call it every time right?

File: src/VisualStudio/VisualBasic/Impl/Options/AdvancedOptionPageControl.xaml.vb (hunk 3 above)
Comment (akhera99) [reply src/VisualStudio/VisualBasic/Impl/Options/AdvancedOptionPageControl.xaml.vb]: misclicked, no confusion. changed

**Pull request closed without merging**